src/rgw/rgw_kms.cc

```diff
@@ -0,0 +1,275 @@
+                                      boost::string_view key_id,
+                                      bufferlist *secret_bl)
+{
+  // TODO: implement
+  return -1;
+}
+
+int request_key_from_vault_with_token(CephContext *cct,
+                                      boost::string_view key_id,
+                                      bufferlist *secret_bl)
+{
+  // These conf settings are static strings so they're read only once from disk
+  static std::string token_file, vault_url;
+  // The actual Vault token is read from file on every request
+  std::string vault_token;
+  int res = 0;
+
+  token_file = cct->_conf->rgw_crypt_s3_kms_vault_token_file;
+  if (token_file.empty()) {
+    ldout(cct, 0) << "ERROR: Vault token file is not set" << dendl;
+    return -EINVAL;
+  }
+  ldout(cct, 20) << "Vault token file: " << token_file << dendl;
+
+  char buf[2048];
+  res = safe_read_file("", token_file.c_str(), buf, sizeof(buf));
+  if (res < 0) {
+    if (-ENOENT == res) {
+      ldout(cct, 0) << "ERROR: Token file '" << token_file << "' not found  " << dendl;
+    } else if (-EACCES == res) {
+      ldout(cct, 0) << "ERROR: Permission denied reading token file" << dendl;
+    } else {
+      ldout(cct, 0) << "ERROR: Failed to read token file with error " << res << dendl;
+    }
+    return res;
+  }
+  // drop trailing newlines
+  while (res && isspace(buf[res-1])) {
+    --res;
+  }
+  vault_token = std::string{buf, static_cast<size_t>(res)};
+
+  vault_url = cct->_conf->rgw_crypt_s3_kms_vault_url;
+  if (vault_url.empty()) {
+    ldout(cct, 0) << "ERROR: Vault URL is not set" << dendl;
+    return -EINVAL;
+  }
+
+  std::string secret_url = vault_url + std::string(key_id);
+  RGWHTTPTransceiver secret_req(cct, "GET", secret_url, secret_bl);
+  secret_req.append_header("X-Vault-Token", vault_token);
+  res = secret_req.process(null_yield);
+  if (res <= 0) {
+    ldout(cct, 0) << "ERROR: Request to Vault failed with error " << res << dendl;
+    return res;
+  }
+
+  if (secret_req.get_http_status() ==
+      RGWHTTPTransceiver::HTTP_STATUS_UNAUTHORIZED) {
+    ldout(cct, 0) << "ERROR: Vault request failed authorization" << dendl;
+    return -EACCES;
+  }
+
+  ldout(cct, 20) << "Request to Vault returned " << res << " and HTTP status "
+    << secret_req.get_http_status() << dendl;
+  return res;
+}
+
+int get_actual_key_from_vault(CephContext *cct,
+                              boost::string_view key_id,
+                              std::string& actual_key)
+{
+  int res = 0;
+  std::string auth;
+  bufferlist secret_bl;
+
+  auth = cct->_conf->rgw_crypt_s3_kms_vault_auth;
+  ldout(cct, 20) << "Vault auhentication method " << auth << dendl;
+
+  if ("token" == auth) {
+    res = request_key_from_vault_with_token(cct, key_id, &secret_bl);
+  } else if ("agent" == auth) {
+    res = request_key_from_vault_with_agent(cct, key_id, &secret_bl);
+  } else {
+    ldout(cct, 0) << "ERROR: Unsupported authentication method " << auth << dendl;
```

Comment:
error message should include the name of rgw_crypt_s3_kms_vault_auth

Reply:
Will do